source/elements/oneTBB/source/algorithms/blocked_ranges/blocked_rangeNd_cls.rst

```diff
@@ -0,0 +1,153 @@
+.. SPDX-FileCopyrightText: 2019-2024 Intel Corporation
+..
+.. SPDX-License-Identifier: CC-BY-4.0
+
+===============
+blocked_range3d
+===============
+**[algorithms.blocked_range3d]**
+
+Class template that represents a recursively divisible N-dimensional half-open interval.
+
+A ``blocked_rangeNd`` is the N-dimensional extension of ``blocked_range``, ``blocked_range2d`` and ``blocked_range3d``.
+We can interpret it as a tensor-product of N instances of ``blocked_range``.
+Different to ``blocked_range2d`` and ``blocked_range3d``, all ranges have to be specified over the same type Value.
+Therefore, blocked_rangeNd<int,2> is a specialisation of blocked_range2d where both dimensions have int type.
+
+.. code:: cpp
+
+    namespace oneapi {
+        namespace tbb {
+            template<typename Value, unsigned int N, typename = detail::make_index_sequence<N>>
+            class blocked_rangeNd {
+            public:
+                // Types
+                using value_type = Value;
+                using dim_range_type = tbb::blocked_range<Value>;
+
+                // Constructors
+                blocked_rangeNd(const dim_range_type<Is>&... args);
+                blocked_rangeNd(
+                  Value begin[N],
```

Comment:
Is there a reason to prefer blocked_rangeNd<int,4> r( {1,2,3,4}, {5,6,7,8}, 4 ) over  blocked_rangeNd<int,4> r( {1, 5}, {2, 6}, {3, 7}, {4, 8}, 4 )?

Reply:
Note that the constructor at line 29 can also work with braced init lists, like this: `blocked_rangeNd<int,4> r{ {1,5,4}, {2,6,4}, {3,7,4}, {4,8,4}};` if the grainsize is explicitly specified as 4.

Reply:
I have looked at the use of blocked_rangeNd in Peano (https://gitlab.lrz.de/hpcsoftware/Peano/-/blob/p4/src/peano4/utils/tbb/Loop.h), and I think we can define two constructors from arrays:
```
blocked_rangeNd( Value begin[N], Value end[N], size_type grainsize[N] );
blocked_rangeNd( Value size[N], size_type grainsize = 1 ); // assumes all dimensions are {0, size[i]}
```
The first one is the alternative way to specify dimensions in "structure of arrays"-like manner, and the second is the convenient helper to both set identical grainsize **and** set all dimensions to start from 0. This one is what Peano would use.

While constructors that do only one of the two simplifications are also possible, I am not sure if they are worth having. Also we can add more syntax sugar later if justification appears.

What do you think?

Reply:
I think the first doesn't offer much over the braced init lists already supported by the constructor at line 29. But then I prefer the begin & end to be near each other for readability. The second suggested constructor seems very reasonable since its I'd suppose that it is the most common case.

Reply:
It's OK to go with only the most simplified version that omits begins and uses a single grainsize. Having a general way to construct an ND range from arrays (as an alternative to a set of 1D ranges) might in principle be meaningful, but if you think it doesn't add much value, I'm fine without it.

Reply:
I think that's reasonable. I honestly never saw a real case for different grain sizes in the different dimensions unless(!) the innermost where you wanna vectorise. But then I'm not sure how good the compiler supports that kind of multilevel parallelisation (vectorisation inside, shared mem outside) anyway. So let's stick with one grain size instead of a set.

Reply:
I implemented what I have understood to be the consensus here in my new commit: The two main constructors whould be `blocked_rangeNd(const dim_range_type<Is>&... args);` and `blocked_rangeNd( Value size[N], size_type grainsize = 1 );`. Please point out in case I misunderstood the discussion and the original `blocked_rangeNd( Value begin[N], Value end[N], size_type grainsize = 1 );` is still needed.

Reply:
> In the suggested version grainsize is set to 1 by default. Is this the desired way?

It matches the default grainsize value for all other blocked ranges.